### PR TITLE
Setup e2e tests on a single chain; add balances query test

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #1124

## Description

The package e2e defines an integration testing suite used for full end-to-end
testing functionality.

The file e2e_suite_test.go defines the testing suite and contains the core
bootstrapping logic that creates a testing environment via Docker containers.
A testing network is created dynamically with 2 test validators.

The file e2e_test.go contains the actual end-to-end integration tests that
utilize the testing suite.

Currently, there is a single test in `e2e_test.go` to query the balances of a validator.

More tests are to be added in future PRs

## Test Run
Proof of successful e2e test run: https://github.com/p0mvn/osmosis/runs/5830443179?check_suite_focus=true

## Next Steps

The following can be split into separate tasks / PRs and tackled one by one:

- run a second chain by reusing the current setup
- add ibc relayer functionality and enable ibc tests
- port all tests from:
   * https://github.com/cosmos/gaia/tree/main/tests/e2e
   * https://github.com/umee-network/umee/tree/main/tests/e2e
- identify more e2e tests specific for Osmosis and write up issues to add them

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ jobs:
       - name: Display go version
         run: go version
       - name: Run all tests
-        run: go test -mod=readonly -timeout 30m -coverprofile=coverage.txt -tags='norace' -covermode=atomic `go list ./... | grep -v simapp`
+        run: go test -mod=readonly -timeout 30m -coverprofile=coverage.txt -tags='norace' -covermode=atomic `go list ./... | grep -E -v 'simapp|e2e'`
```
Comment (alexanderbez) [inline .github/workflows/test.yml:40]: nit: we should instead encapsulate this in a make target.

See the changes I made here: https://github.com/cosmos/gaia/blob/main/Makefile#L167-L205

I don't think it's too much trouble to do that in this PR. WDYT?

File: go.mod [hunk 2]
```diff
@@ -29,6 +29,27 @@ require (
 	gopkg.in/yaml.v2 v2.4.0
 )
 
+require (
```
Comment (alexanderbez) [inline go.mod:32]: Did you run `go mod tidy --compat=1.17`? Weird, these should be bundled with the existing `indirect` require block below.

File: tests/e2e/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# End-to-end Tests
```
Comment (alexanderbez) [inline tests/e2e/README.md:1]: Don't worry @ValarDragon I'll beef up this doc after merge.

**Review by alexanderbez — APPROVED** [review]
🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Comment by nikever** [comment]
Overall LGTM 🚀

I'm actually reverting to the previous `Dockerfile` to avoid these kind of permissions issues so we don't have to create a new `e2e.Dockerfile`

**Comment by alexanderbez** [comment]
> Overall LGTM 🚀
> 
> I'm actually reverting to the previous `Dockerfile` to avoid these kind of permissions issues so we don't have to create a new `e2e.Dockerfile`

So the context on why I originally named it `e2e.Dockerfile` is because before it was just `Dockerfile` and validators/node operators _assumed_ that was the image they should use for mainnet, which it wasn't because it has (or can have) specific E2E stuff in it, not intended for mainnet.

**Comment by nikever** [comment]
> So the context on why I originally named it `e2e.Dockerfile` is because before it was just `Dockerfile` and validators/node operators _assumed_ that was the image they should use for mainnet, which it wasn't because it has (or can have) specific E2E stuff in it, not intended for mainnet.

That makes absolutely sense but in our setup we are just running a _working_ copy of the `Dockerfile` that currently has permission problems if not run correctly.  

Ideally we should try to use the same `Dockerfile` for `prod` and `e2e` but if they will diverge in the future I'm happy with your solution.

File: go.mod (hunk 2 above)
Comment (ValarDragon) [reply go.mod:32]: Yeah, definitely looks like something went not-quite-right here

File: tests/e2e/e2e_setup_test.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+	genFilePath := config.GenesisFile()
+	appGenState, genDoc, err := genutiltypes.GenesisStateFromGenFile(genFilePath)
+	s.Require().NoError(err)
+
+	var bankGenState banktypes.GenesisState
+	s.Require().NoError(cdc.UnmarshalJSON(appGenState[banktypes.ModuleName], &bankGenState))
+
+	bankGenState.DenomMetadata = append(bankGenState.DenomMetadata, banktypes.Metadata{
+		Description: "An example stable token",
```
Comment (ValarDragon) [inline tests/e2e/e2e_setup_test.go:147]: Oh wow, I didn't know there was a description field available in the bank module

**Review by ValarDragon — APPROVED** [review]
Didn't review scaffolding in depth. But the test looks great / its pretty exciting if it 'justworks.TM'.

Should we do the makefile change in this PR or a separate one?

**Comment by p0mvn** [comment]
Making the makefile change here and then will merge

File: go.mod (hunk 2 above)
Comment (p0mvn) [reply go.mod]: Had to manually move them to the same block and run `go mod tidy`

**Comment by p0mvn** [comment]
On another look, I think the makefile should be addressed in a separate PR. Otherwise, the scope ends up being too large. I will merge this and address it in a separate PR

**Pull request merged**